src/main/java/io/github/thebusybiscuit/slimefun4/core/services/MetricsService.java

```diff
@@ -0,0 +1,237 @@
+package io.github.thebusybiscuit.slimefun4.core.services;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import io.github.thebusybiscuit.slimefun4.utils.PatternUtils;
+import kong.unirest.HttpResponse;
+import kong.unirest.JsonNode;
+import kong.unirest.Unirest;
+import kong.unirest.UnirestException;
+import me.mrCookieSlime.Slimefun.api.Slimefun;
+import org.bukkit.plugin.Plugin;
+
+/**
+ * This Class represents a Metrics Service that sends data to https://bstats.org/
+ * This data is used to analyse the usage of this {@link Plugin}.
+ * <p>
+ * You can find more info in the README file of this Project on GitHub. <br>
+ * <b>Note:</b> To start the metrics you will need to be calling {@link #start()}
+ *
+ * @author TheBusyBiscuit
+ * @author WalshyDev
+ */
+public class MetricsService {
+
+    private static final String REPO_NAME = "MetricsModule";
+    private static final String GH_API = "https://api.github.com/repos/Slimefun/" + REPO_NAME;
+    private static final String GH_REPO_RELEASES = "https://github.com/Slimefun/" + REPO_NAME
+        + "/releases/download";
+
+    private final SlimefunPlugin plugin;
+    private final File metricFile;
+
+    private URLClassLoader moduleClassLoader;
+    private String metricVersion = null;
+    private boolean newlyDownloaded = false;
+
+    static {
+        Unirest.config()
+            .concurrency(2, 1)
+            .setDefaultHeader("User-Agent", "MetricsModule Auto-Updater")
+            .setDefaultHeader("Accept", "application/vnd.github.v3+json")
+            .enableCookieManagement(false)
+            .cookieSpec("ignoreCookies");
+    }
+
+    public MetricsService(SlimefunPlugin plugin) {
+        this.plugin = plugin;
+        this.metricFile = new File(plugin.getDataFolder(), REPO_NAME + ".jar");
```

Comment:
Also about this location, we have a cache subfolder in the data folder for all the github files, should we move it there? Then they wont be bothered with this jar sitting inbetween their configs?